### PR TITLE
Add localisation for overlay headers

### PR DESCRIPTION
**Pull request by kj415j45**
Depends on:
- [x] #13911
- [x] ppy/osu-web#7937
- [x] osu-resources strings update and bump

### CODE REVIEW

**Comment by bdach** [comment]
Please don't open incomplete PRs. If you wish to reserve a task to work on, open an issue for it instead and note that you will be working on it.

File: osu.Game/Overlays/News/NewsHeader.cs [hunk 1]
```diff
@@ -4,20 +4,24 @@
...
+using osu.Game.Localisation;
 
 namespace osu.Game.Overlays.News
 {
     public class NewsHeader : BreadcrumbControlOverlayHeader
     {
-        private const string front_page_string = "frontpage";
+        public static LocalisableString FrontPageString => NewsStrings.FrontPageString;
+        public static LocalisableString HeaderTitle => NewsStrings.HeaderTitle;
+        public static LocalisableString HeaderDescription => NewsStrings.HeaderDescription;
```
Comment (bdach) [inline osu.Game/Overlays/News/NewsHeader.cs:16]: Should be `public static readonly`.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
There should be no `osu-resources` changes required, as I understand the strings being translated are already translated on `osu-web`. Those are [already available for consumption](https://github.com/ppy/osu-resources/tree/master/osu.Game.Resources/Localisation/Web).

**Comment by kj415j45** [comment]
> There should be no `osu-resources` changes required (except for maybe `NewsStrings.HeaderDescription`), as I understand the strings being translated are already translated on `osu-web`. Those are [already available for consumption](https://github.com/ppy/osu-resources/tree/master/osu.Game.Resources/Localisation/Web?rgh-link-date=2021-07-17T12%3A22%3A17Z).

Will check if there're available strings in web later

**Comment by peppy** [comment]
As already said, please remove the remaining tasks from this PR and do them separately.

File: osu.Game/Overlays/News/NewsHeader.cs (hunk 1 above)
Comment (kj415j45) [reply osu.Game/Overlays/News/NewsHeader.cs:16]: `readonly` is invalid for this.
this code format is the same as other localisation PR.

Comment (bdach) [reply osu.Game/Overlays/News/NewsHeader.cs:16]: Fair, misread.

File: osu.Game/Overlays/Wiki/WikiHeader.cs [hunk 2]
```diff
@@ -5,14 +5,18 @@
...
 
 namespace osu.Game.Overlays.Wiki
 {
     public class WikiHeader : BreadcrumbControlOverlayHeader
     {
-        private const string index_page_string = "index";
         private const string index_path = "Main_Page";
+        public static LocalisableString IndexPageString => WikiStrings.IndexPageString;
+        public static LocalisableString HeaderTitle => WikiStrings.HeaderTitle;
+        public static LocalisableString HeaderDescription => WikiStrings.HeaderDescription;
```
Comment (frenzibyte) [inline osu.Game/Overlays/Wiki/WikiHeader.cs:19]: I don't really see why this is declared twice, when you could just prepend `WikiStrings` in every usage, making it read even better. (same applies to `NewsHeader`)

Comment (kj415j45) [reply osu.Game/Overlays/Wiki/WikiHeader.cs:19]: Full PR Reference:

[1] [LocalisableString usage](https://github.com/peppy/osu/blob/master/osu.Game/Overlays/ChatOverlay.cs#L38-L39)
```csharp
        public LocalisableString Title => ChatStrings.HeaderTitle;
        public LocalisableString Description => ChatStrings.HeaderDescription;
```

[2] [Code format of `*Strings.cs`](https://github.com/ppy/osu/pull/12495#issue-619104156)
```csharp
/// <summary>
/// "Fallback / default text"
/// </summary>
public static LocalisableString LookupKey => new TranslatableString(getKey("lookup_key"), "Fallback / default text");
```

p.s. will remove `static` from `*Header.cs` to match [1]

Comment (bdach) [reply osu.Game/Overlays/Wiki/WikiHeader.cs:19]: The difference there is that `ChatOverlay` was *forced* to redeclare those members, because of implementing `INamedOverlayComponent`.

That said this PR has several places where the same string is referred to multiple times, so I'm not overly fussed about this myself.

Comment (kj415j45) [reply osu.Game/Overlays/Wiki/WikiHeader.cs:19]: I see.

File: osu.Game/Overlays/News/NewsHeader.cs [hunk 3]
```diff
@@ -4,20 +4,22 @@
...
+using osu.Game.Localisation;
+using osu.Game.Resources.Localisation.Web;
 
 namespace osu.Game.Overlays.News
 {
     public class NewsHeader : BreadcrumbControlOverlayHeader
     {
-        private const string front_page_string = "frontpage";
-
+        public LocalisableString FrontPageString => osu.Game.Resources.Localisation.Web.NewsStrings.IndexTitleInfo;
```
Comment (frenzibyte) [inline osu.Game/Overlays/News/NewsHeader.cs:15]: should be static, also should have a newline between a property and a field.
```suggestion
        public static LocalisableString FrontPageString => osu.Game.Resources.Localisation.Web.NewsStrings.IndexTitleInfo;

```

File: osu.Game/Overlays/Wiki/WikiHeader.cs [hunk 4]
```diff
@@ -5,14 +5,16 @@
...
 using osu.Game.Online.API.Requests.Responses;
+using osu.Game.Resources.Localisation.Web;
 
 namespace osu.Game.Overlays.Wiki
 {
     public class WikiHeader : BreadcrumbControlOverlayHeader
     {
-        private const string index_page_string = "index";
         private const string index_path = "Main_Page";
+        public LocalisableString IndexPageString => LayoutStrings.HeaderHelpIndex;
```
Comment (frenzibyte) [inline osu.Game/Overlays/Wiki/WikiHeader.cs:17]: static + newline between field and property.
```suggestion

        public static LocalisableString IndexPageString => LayoutStrings.HeaderHelpIndex;
```

**Review by frenzibyte — CHANGES_REQUESTED** [review]
If you're planning for full overlay header localisation, `BeatmapSetOverlay`, `ChatOverlay`, and `UserProfileOverlay` should be localised as well.

While I was checking with osu!web, the strings used for the overlay titles are all stored on `PageTitleStrings`, but some strings used here don't look from there:
https://github.com/ppy/osu/blob/f80b18377775e3bf52b4370d0bd7459dc383112c/osu.Game/Overlays/Changelog/ChangelogHeader.cs#L120

https://github.com/ppy/osu/blob/f80b18377775e3bf52b4370d0bd7459dc383112c/osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs#L36

These use the strings from the web's navigation menu items, displaying as the overlay's title on the header.

I presume that this is done to look theoretically correct on the toolbar when hovering over the overlay buttons (as a somewhat analogue to web's navigation menu), but that's weird for overlay header titles, and has already caused an issue which has been dealt with weirdly:

https://github.com/ppy/osu/blob/1b4bff0d9fa52dec5b18f33158b4201bcb1ed2e2/osu.Game/Overlays/Wiki/WikiHeader.cs#L78-L79

(now the overlay title uses web's navigation menu item text, and the overlay description uses the page title which is supposed to be for the overlay title..??)

I think what should happen here is adding a separate `ShortTitle` property on `INamedOverlayComponent`s for the toolbar to use instead, and let `WikiOverlay` set that to `LayoutStrings.MenuHelpGetWiki`. 

Then let all overlays use the correct `PageTitleStrings` for `Title`.

File: osu.Game/Localisation/HeaderDescriptionStrings.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Localisation;
+
+namespace osu.Game.Localisation
+{
+    public static class HeaderDescriptionStrings
```
[the remark this answers is not in the document]
Comment (kj415j45) [reply osu.Game/Localisation/HeaderDescriptionStrings.cs:8]:
```csharp
        /// <inheritdoc cref="PageTitleStrings.MainBeatmapsetsControllerIndex"/>
        public static LocalisableString BeatmapListingTitle => PageTitleStrings.MainBeatmapsetsControllerIndex;
```
Confirmed that this doesn't work on current version of osu-localisation-analyser. The reason is missing `getKey()` method fot it to identify the key properly.

[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Localisation/HeaderDescriptionStrings.cs:8]: It's supposed to just be ignored by the generator, not sure what you mean?

**Comment by kj415j45** [comment]
Have to admit that I do mis-match string of `Changelog`'s title.

---

But the `Ranking(s)`' title is another tiny problem:
`LayoutStrings.MenuRankingsDefault` match current in-game string `rankings`. however, `PageTitleStrings.MainRankingControllerDefault` which is suppose to be used is `ranking`, missing an `s`.

**Comment by frenzibyte** [comment]
Not sure why they're different, but `PageTitleStrings.MainRankingControllerDefault` will be used for the rankings overlay header title, and `PageTitleStrings.MainRankingControllerDefault` will be used for the toolbar name, after I get to adding `ShortTitle` to `INamedOverlayComponent` (or a better name because it doesn't make sense now).

File: osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs [hunk 6]
```diff
@@ -29,18 +33,44 @@ private class RankingsTitle : OverlayTitle
...
+                    return LayoutStrings.MenuRankingsScore;
+
+                case RankingsScope.Country:
+                    return LayoutStrings.MenuRankingsCountry;
+
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(value), value, null);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs:74]: Not sure how valid this is. Existing localised enums have localisations in the same file.

**Comment by bdach** [comment]
I'm not sure why the scope of this was expanded to all overlays, going against the advice [given above](https://github.com/ppy/osu/pull/13914#issuecomment-881896001)...

File: osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs:74]: Probably failed on the wording, but I meant that `RankingsScope` and `RankingsScopeEnumLocalisationMapper` should both move to one `RankingsScope.cs` file.

**Comment by frenzibyte** [comment]
I failed to notice that comment as the author started to localise everything in this PR, but in the end, I think it's not that complex, it's just a couple of strings in each overlay's header pointing to the right `LocalisableString` in the `Strings` files.

**Comment by bdach** [comment]
This needs conflict resolution now.

File: osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kj415j45) [reply osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs:74]: This change will go to another PR. won't fix in this one.

[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs]: It's a pretty straightforward change so I've done it myself with another fix for the used strings in the mapper (https://github.com/ppy/osu/pull/13914/commits/456f4e6f1fce118405e0947f73705e90d1c30306).

**Comment by frenzibyte** [comment]
I've resolved some of the remaining issues I didn't notice on my first review, I think it's in a good state now.

I was hoping the existing title and description strings in `SettingsStrings`/`NotificationsStrings`/etc. to be moved to the new `NamedOverlayComponentStrings`, but very hesitant about doing that now that they're already published to crowdin, so I've left them as they're for now.

This will have to wait for https://github.com/ppy/osu/pull/13918 to conclude, however.

**Comment by peppy** [comment]
This needs conflict resolution too.

File: osu.Game/Localisation/NamedOverlayComponentStrings.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        /// <inheritdoc cref="PageTitleStrings.MainNewsControllerDefault"/>
+        public static LocalisableString NewsTitle => PageTitleStrings.MainNewsControllerDefault;
+
+        /// <summary>
+        /// "get up-to-date on community happenings"
+        /// </summary>
+        public static LocalisableString NewsDescription => new TranslatableString(getKey(@"news_description"), @"get up-to-date on community happenings");
+
+        /// <inheritdoc cref="PageTitleStrings.MainWikiControllerDefault"/>
+        public static LocalisableString WikiTitle => PageTitleStrings.MainWikiControllerDefault;
```
Comment (peppy) [inline osu.Game/Localisation/NamedOverlayComponentStrings.cs:57]: Why are these here? Just use the osu-web ones directly, I think?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Localisation/NamedOverlayComponentStrings.cs (hunk 7 above)
Comment (kj415j45) [reply osu.Game/Localisation/NamedOverlayComponentStrings.cs:57]: `PageTitleStrings` comes from `osu.Game.Resources.Localisation.Web` namespace.

---

and as discussed before, the `*Title` and `*Description` style is for consistency:
```cs
//in components
{
    Title = NamedOverlayComponentStrings./* name here */Title;
    Description = NamedOverlayComponentStrings./* name here */Description;
}
//in osu.Game/Localisation/NamedOverlayComponentStrings.cs
    /// <inheritdoc cref="PageTitleStrings.{string name}"/>
    public static LocalisableString /* name here */Title => PageTitleStrings./* name here */ControllerDefault;

    /// <summary>
    /// "sample string"
    /// </summary>
    public static LocalisableString /* name here */Description => new TranslatableString(getKey(@"_description"), @"sample string");
```

But it's ok using following style :shrug:
```cs
//in components
{
    Title = PageTitleStrings./* name here */ControllerDefault;
    Description = NamedOverlayComponentStrings./* name here */Description;
}
```

Comment (peppy) [reply osu.Game/Localisation/NamedOverlayComponentStrings.cs:57]: I definitely prefer the direct references. @smoogipoo can you confirm your opinion?

Comment (smoogipoo) [reply osu.Game/Localisation/NamedOverlayComponentStrings.cs:57]: I also prefer direct references.

Comment (kj415j45) [reply osu.Game/Localisation/NamedOverlayComponentStrings.cs:57]: 👌🏼 it will be direct references in next commit

Comment (peppy) [reply osu.Game/Localisation/NamedOverlayComponentStrings.cs:57]: Not sure if it will ever be the case, but I can see these potentially being moved to `osu-web` in the future (if there are more design updates to bring web/client in closer consistency).

File: osu.Game/Overlays/Changelog/ChangelogHeader.cs [hunk 8]
```diff
@@ -21,16 +24,16 @@ public class ChangelogHeader : BreadcrumbControlOverlayHeader
 
         public ChangelogUpdateStreamControl Streams;
 
-        private const string listing_string = "listing";
+        public static LocalisableString ListingString => LayoutStrings.HeaderChangelogIndex;
```
Comment (peppy) [inline osu.Game/Overlays/Changelog/ChangelogHeader.cs:27]: @smoogipoo thoughts on this one? is used in tests (`public` part of the weirdness) but not sure if we want this, or another case of just inlining the references at each point.

File: osu.Game/Overlays/Changelog/ChangelogHeader.cs [hunk 9]
```diff
@@ -21,16 +24,16 @@ public class ChangelogHeader : BreadcrumbControlOverlayHeader
...
         private Box streamsBackground;
 
         public ChangelogHeader()
         {
-            TabControl.AddItem(listing_string);
+            TabControl.AddItem(ListingString);
             Current.ValueChanged += e =>
             {
-                if (e.NewValue == listing_string)
+                if (e.NewValue == ListingString)
```
Comment (peppy) [inline osu.Game/Overlays/Changelog/ChangelogHeader.cs:36]: Genuinely surprised this actually works across language changes..

File: osu.Game/Overlays/Changelog/ChangelogHeader.cs (hunk 8 above)
Comment (smoogipoo) [reply osu.Game/Overlays/Changelog/ChangelogHeader.cs:27]: I think I'm fine with this. I just don't like redirects in the supposed-to-be-autogenerated XStrings classes because the format isn't guaranteed to remain when new strings are added and I can't guarantee it won't break the analyser in the first place.

But this seems okay to me.

**Review by peppy — APPROVED** [review]

File: osu.Game/Localisation/NamedOverlayComponentStrings.cs (hunk 7 above)
Comment (frenzibyte) [reply osu.Game/Localisation/NamedOverlayComponentStrings.cs]: Yeah consistency is really the only reason I've went with that style, but since redirecting in generated strings classes is not a good idea in general, I'm fine with having it use direct references.

Moving the tooltip descriptions web-side would also remove all this hassle and leave things in a consistent state, but that's probably for later when actually needed.
